## 1. Μεταβλητές και τύποι
Μια μεταβλητή είναι ένα όνομα που χρησιμοποιούμε για να αναφερθούμε σε μια τιμή. Το `=` αναθέτει την τιμή στα δεξιά στο όνομα στα αριστερά.

Στο παράδειγμα, κάθε τιμή περιγράφει μια **υποθετική επιστολή**. Η `print()` εμφανίζει τιμές και η `type()` δείχνει τον τύπο τους.

Πριν εκτελέσεις: ποια μεταβλητή είναι αριθμός και ποια κείμενο;

In [1]:
year = 1880
place = "Αθήνα"
words = 120
is_digitised = True

print(place)
print(words)
print(type(year))
print(type(place))
print(type(is_digitised))

Αθήνα
120
<class 'int'>
<class 'str'>
<class 'bool'>


**Αναμενόμενα:** `Αθήνα`, `120`, και οι τύποι `int`, `str`, `bool`.

| Τύπος | Παράδειγμα | Χρήση |
|---|---|---|
| `int` | `120` | Ακέραιος |
| `float` | `12.5` | Αριθμός κινητής υποδιαστολής |
| `str` | `"Αθήνα"` | Κείμενο |
| `bool` | `True` | Λογική τιμή |

Στην Python οι δεκαδικοί γράφονται με **τελεία**. Τα εισαγωγικά δηλώνουν κείμενο. Τα κεφαλαία και τα πεζά στα ονόματα διαφέρουν.

In [2]:
print(120 + 30)
print("120" + "30")

words_text = "120"
converted_words = int(words_text)
print(converted_words + 30)

150
12030
150


**Αναμενόμενη έξοδος:** `150`, `12030`, `150` σε τρεις γραμμές.

Η δεύτερη πράξη ενώνει κείμενα. Η `int("120")` μετατρέπει κατάλληλο κείμενο σε ακέραιο. Η μετατροπή `int("άγνωστο")` θα προκαλούσε σφάλμα· δεν την εκτελούμε εδώ.

### Άσκηση 1 · Ένα ποσοστό
Έχουμε 5 επιστολές από την Αθήνα σε σύνολο 12. Συμπλήρωσε τις τιμές και αντικατάστησε το `None` με τον υπολογισμό του ποσοστού.

In [3]:
# TODO: Συμπλήρωσε τα πλήθη και τον υπολογισμό.
athens_count = 0
total_count = 0
athens_percentage = None

print("Ποσοστό Αθήνας:", athens_percentage)

Ποσοστό Αθήνας: None


## 2. Λίστες και υπολογισμοί
Η λίστα κρατά πολλές τιμές με συγκεκριμένη σειρά. Τα στοιχεία γράφονται σε αγκύλες και χωρίζονται με κόμματα.

Οι επόμενες έξι τιμές είναι οι αριθμοί λέξεων των πρώτων έξι συνθετικών επιστολών της 1ης εβδομάδας. **Μία τιμή αντιστοιχεί σε μία επιστολή.**

In [4]:
word_counts = [120, 180, 90, 210, 150, 130]

print("Λίστα:", word_counts)
print("Πλήθος επιστολών:", len(word_counts))
print("Πρώτη τιμή:", word_counts[0])
print("Τρίτη τιμή:", word_counts[2])
print("Τελευταία τιμή:", word_counts[-1])
print("Τμήμα θέσεων 1 έως 3:", word_counts[1:4])

Λίστα: [120, 180, 90, 210, 150, 130]
Πλήθος επιστολών: 6
Πρώτη τιμή: 120
Τρίτη τιμή: 90
Τελευταία τιμή: 130
Τμήμα θέσεων 1 έως 3: [180, 90, 210]


**Έλεγξε:** πλήθος `6`, πρώτη τιμή `120`, τρίτη `90`, τελευταία `130`, τμήμα `[180, 90, 210]`.

Η πρώτη θέση έχει δείκτη **0**. Στο `[1:4]`, η θέση 4 **δεν περιλαμβάνεται**. Η επιλογή τμήματος δεν διαγράφει στοιχεία από το αρχικό `word_counts`.

Η `len()` μετρά στοιχεία, ενώ η `sum()` προσθέτει τις τιμές τους.

In [5]:
total_words = sum(word_counts)
number_of_letters = len(word_counts)
mean_words = total_words / number_of_letters

print("Σύνολο λέξεων:", total_words)
print("Μέση τιμή:", round(mean_words, 2))

Σύνολο λέξεων: 880
Μέση τιμή: 146.67


**Αναμενόμενα:** `880` λέξεις συνολικά και περίπου `146.67` λέξεις ανά επιστολή. Ο υπολογισμός της μέσης τιμής εδώ προϋποθέτει ότι η λίστα δεν είναι κενή.

### Άσκηση 2 · Επίλεξε και υπολόγισε
Στην παρακάτω λίστα:
- επίλεξε την τρίτη τιμή,
- επίλεξε τις τρεις πρώτες τιμές ως νέα λίστα,
- υπολόγισε το άθροισμα όλων των τιμών.

Κάθε τιμή είναι διάρκεια υποθετικής επίσκεψης, σε λεπτά.

In [7]:
durations_ex2 = [40, 20, 60, 30]

# TODO: Αντικατάστησε τα None με εκφράσεις που χρησιμοποιούν τη λίστα.
third_duration = None
first_three_durations = None
total_duration = None

print("Τρίτη τιμή:", third_duration)
print("Πρώτες τρεις:", first_three_durations)
print("Συνολικά λεπτά:", total_duration)

Τρίτη τιμή: None
Πρώτες τρεις: None
Συνολικά λεπτά: None


### Αλλαγή σε αντίγραφο
Η `append()` προσθέτει στοιχείο στο τέλος. Για να μη μεταβάλουμε την αρχική λίστα αριθμών, δημιουργούμε αντίγραφο με `.copy()`.

Το παράδειγμα αναδημιουργεί το αντίγραφο σε κάθε εκτέλεση, ώστε να μην προσθέτει επανειλημμένα την τιμή 240.

In [6]:
working_counts = word_counts.copy()
working_counts.append(240)

print("Αρχική:", word_counts)
print("Αντίγραφο:", working_counts)
print("Πλήθος αντιγράφου:", len(working_counts))

Αρχική: [120, 180, 90, 210, 150, 130]
Αντίγραφο: [120, 180, 90, 210, 150, 130, 240]
Πλήθος αντιγράφου: 7


**Αναμενόμενο:** η αρχική λίστα έχει 6 τιμές και το αντίγραφο 7. Το `working_counts = word_counts` θα έδινε δύο ονόματα στην ίδια λίστα, χωρίς ανεξάρτητο αντίγραφο.

Η `.copy()` αρκεί για αυτό το παράδειγμα με αριθμούς. Δεν αποτελεί βαθύ αντίγραφο εμφωλευμένων δομών.

## 3. Συνθήκες
Μια σύγκριση δίνει `True` ή `False`. Το `if` χρησιμοποιεί αυτό το αποτέλεσμα για να επιλέξει εντολές.

| Έκφραση | Σημασία |
|---|---|
| `words > 150` | Περισσότερες από 150 λέξεις |
| `words >= 150` | Τουλάχιστον 150 λέξεις |
| `words == 150` | Ακριβώς 150 λέξεις |
| `words < 150` | Λιγότερες από 150 λέξεις |

Το `=` κάνει ανάθεση, ενώ το `==` σύγκριση. Οι εσοχές και η `:` είναι μέρος της σύνταξης.

In [7]:
example_words = 150

if example_words >= 150:
    print("Τουλάχιστον 150 λέξεις")
else:
    print("Κάτω από 150 λέξεις")

Τουλάχιστον 150 λέξεις


**Αναμενόμενο:** `Τουλάχιστον 150 λέξεις`.


### Άσκηση 3 · Διόρθωσε τον κανόνα
Μια επίσκεψη θεωρείται επιλεγμένη για την άσκηση όταν διαρκεί **τουλάχιστον 30 λεπτά**. Ο παρακάτω κώδικας έχει λογικό σφάλμα. Άλλαξε μόνο τον τελεστή της συνθήκης.

In [8]:
visit_minutes_ex3 = 30

# TODO: Διόρθωσε τη συνθήκη ώστε να περιλαμβάνει και το 30.
if visit_minutes_ex3 > 30:
    selected_ex3 = True
else:
    selected_ex3 = False

print("Επιλέχθηκε:", selected_ex3)

Επιλέχθηκε: False


**Δοκιμές που θα κάνεις:**

| Διάρκεια | Αναμενόμενη επιλογή |
|---|---|
| 29 | `False` |
| 30 | `True` |
| 31 | `True` |

Επανέφερε την τιμή στο **30** πριν από τον τελικό έλεγχο. Ένας μόνο επιτυχής έλεγχος δεν αποδεικνύει ότι ο κανόνας λειτουργεί για όλες τις εισόδους.

## 4. Επαναλήψεις
Το `for` παίρνει διαδοχικά κάθε στοιχείο της λίστας. Οι εντολές με εσοχή εκτελούνται για κάθε στοιχείο.

In [13]:
word_counts = [120, 180, 90, 210, 150, 130]

for current_words in word_counts:
    print(current_words)

120
180
90
210
150
130


Η έξοδος έχει έξι γραμμές, μία για κάθε τιμή. Το `current_words` παίρνει πρώτα το 120, μετά το 180 και συνεχίζει μέχρι το 130.

### Παράδειγμα · Πόσες επιστολές έχουν τουλάχιστον 150 λέξεις;
Μηδενίζουμε τον μετρητή **πριν** από το `for`. Για κάθε τιμή που ικανοποιεί τη συνθήκη, τον αυξάνουμε κατά ένα.

In [14]:
count_at_least_150 = 0

for current_words in word_counts:
    if current_words >= 150:
        count_at_least_150 = count_at_least_150 + 1

print("Επιστολές με τουλάχιστον 150 λέξεις:", count_at_least_150)

Επιστολές με τουλάχιστον 150 λέξεις: 3


**Αναμενόμενο:** `3`, για τις τιμές 180, 210 και 150.

Αν θέλουμε τις ίδιες τις επιλεγμένες τιμές, δημιουργούμε νέα λίστα αντί για έναν μόνο μετρητή.

In [15]:
selected_word_counts = []

for current_words in word_counts:
    if current_words >= 150:
        selected_word_counts.append(current_words)

print(selected_word_counts)

[180, 210, 150]


**Αναμενόμενο:** `[180, 210, 150]`.

### Άσκηση 4 · Άλλαξε την ερώτηση
Συμπλήρωσε το σώμα της επανάληψης για να μετρήσεις τις επιστολές με **λιγότερες από 150 λέξεις**.

Το `pass` είναι προσωρινή εντολή που δεν κάνει τίποτα. Σου επιτρέπει να εκτελέσεις το αρχικό κελί χωρίς συντακτικό σφάλμα. Αντικατάστησέ το με τον έλεγχο και την αύξηση του μετρητή.

In [16]:
word_counts_ex4 = [120, 180, 90, 210, 150, 130]
count_below_150 = 0

for current_words in word_counts_ex4:
    # TODO: Αντικατάστησε το pass με if και αύξηση του μετρητή.
    pass

print("Λιγότερες από 150 λέξεις:", count_below_150)

Λιγότερες από 150 λέξεις: 0


### Άσκηση 5 · Κράτησε τις τιμές
Δημιούργησε λίστα με τις επιστολές που έχουν **τουλάχιστον 180 λέξεις**. Συμπλήρωσε το σώμα του `for`.

**Αναμενόμενο αποτέλεσμα:** `[180, 210]`. Το αρχικό `word_counts_ex5` δεν πρέπει να αλλάξει.

In [18]:
word_counts_ex5 = [120, 180, 90, 210, 150, 130]
selected_at_least_180 = []

for current_words in word_counts_ex5:
    # TODO: Πρόσθεσε την τιμή στη νέα λίστα όταν ικανοποιεί το όριο.
    pass

print(selected_at_least_180)

[]


## 5. Αφαίρεση περιττών πεδίων
**Σενάριο:** θέλουμε να μετρήσουμε πόσες επισκέψεις σε ένα υποθετικό μουσείο διαρκούν τουλάχιστον 30 λεπτά.

Η διαθέσιμη λίστα περιέχει ετικέτα προσώπου, υποθετικό email, ηλικία και διάρκεια. Το ερώτημά μας χρειάζεται μόνο τη διάρκεια.

Κάθε εσωτερική λίστα είναι μία εγγραφή. Χρησιμοποιούμε την ίδια πρόσβαση με δείκτες που ήδη μάθαμε:

| Θέση | Περιεχόμενο | Αναγκαίο για το ερώτημα; |
|---|---|---|
| 0 | Ετικέτα φανταστικού προσώπου | Όχι |
| 1 | Φανταστικό email | Όχι |
| 2 | Υποθετική ηλικία | Όχι |
| 3 | Διάρκεια επίσκεψης, σε λεπτά | Ναι |

Στο πραγματικό ερευνητικό περιβάλλον, το όνομα, το email και συνδυασμοί άλλων στοιχείων μπορεί να επιτρέπουν αναγνώριση. **Όλα τα παρακάτω στοιχεία είναι φανταστικά.**

In [9]:
fictional_visits = [
    ["Πρόσωπο Α", "a@example.org", 21, 40],
    ["Πρόσωπο Β", "b@example.org", 34, 25],
    ["Πρόσωπο Γ", "c@example.org", 28, 60],
    ["Πρόσωπο Δ", "d@example.org", 45, 30],
    ["Πρόσωπο Ε", "e@example.org", 52, 15],
    ["Πρόσωπο Ζ", "f@example.org", 23, 50]
]

first_record = fictional_visits[0]
print("Πρώτη εγγραφή:", first_record)
print("Διάρκεια πρώτης επίσκεψης:", first_record[3])

Πρώτη εγγραφή: ['Πρόσωπο Α', 'a@example.org', 21, 40]
Διάρκεια πρώτης επίσκεψης: 40


**Αναμενόμενη διάρκεια:** `40` λεπτά.

### Άσκηση 6 · Δημιούργησε αντίγραφο με λιγότερα πεδία
Συμπλήρωσε το σώμα της επανάληψης ώστε το `analysis_minutes` να περιέχει **μόνο τις διάρκειες**, ως αριθμούς.

- Χρησιμοποίησε την `append()`.
- Επίλεξε από κάθε `record` μόνο τη θέση που αντιστοιχεί στη διάρκεια.
- Μην αντιγράψεις ολόκληρη την εγγραφή.
- Μην αλλάξεις το `fictional_visits`.

**Αναμενόμενο:** `[40, 25, 60, 30, 15, 50]`.

In [21]:
analysis_minutes = []

for record in fictional_visits:
    # TODO: Πρόσθεσε μόνο τη διάρκεια επίσκεψης.
    pass

print("Δεδομένα για την ανάλυση:", analysis_minutes)

Δεδομένα για την ανάλυση: []


### Άσκηση 7 · Απάντησε με τα μειωμένα δεδομένα
Χρησιμοποίησε το `analysis_minutes` και μέτρησε τις επισκέψεις διάρκειας **τουλάχιστον 30 λεπτών**.

Πριν εκτελέσεις, γράψε στο επόμενο κελί κειμένου ποιες διάρκειες περιμένεις να επιλεγούν. Αν η άσκηση 6 δεν έχει ολοκληρωθεί, η λίστα θα είναι κενή.

In [23]:
# TODO: Έλεγξε το όριο και αύξησε τον μετρητή.


print("Επισκέψεις τουλάχιστον 30 λεπτών:", )

Επισκέψεις τουλάχιστον 30 λεπτών: 0


## Βοήθεια σε συνηθισμένα προβλήματα

| Πρόβλημα | Τι να ελέγξεις |
|---|---|
| `NameError` | Μήπως δεν εκτελέστηκε η ανάθεση ή άλλαξε η γραφή του ονόματος; |
| `SyntaxError` | Μήπως λείπει εισαγωγικό, παρένθεση ή η `:` μετά το `if`/`for`; |
| `IndentationError` | Έχουν σωστή, συνεπή εσοχή οι εσωτερικές εντολές; |
| `TypeError` | Προσπαθείς να συνδυάσεις ασύμβατους τύπους, όπως αριθμό και κείμενο; |
| `IndexError` | Είναι ο δείκτης μέσα στα όρια της λίστας; |
| Ο μετρητής δίνει 0 ή 1 | Μήπως τον μηδενίζεις μέσα στην επανάληψη; |
| Η λίστα μεγαλώνει σε κάθε εκτέλεση | Μήπως εκτελείς ξανά την `append()` χωρίς νέα αρχικοποίηση; |


## Συνάρτηση με παράμετρο
Η ενότητα αυτή είναι για πρόσθετη εξάσκηση.

Μπορούμε να οργανώσουμε την καταμέτρηση σε συνάρτηση. Οι παράμετροι `values` και `limit` δέχονται τη λίστα και το όριο. Το `return` επιστρέφει το πλήθος.

Το παράδειγμα χρησιμοποιεί δική του λίστα και δεν εξαρτάται από τις απαντήσεις των ασκήσεων.

In [25]:
def count_at_least(values, limit):
    count = 0
    for value in values:
        if value >= limit:
            count = count + 1
    return count

extension_minutes = [40, 25, 60, 30, 15, 50]
print("Τουλάχιστον 30:", count_at_least(extension_minutes, 30))
print("Τουλάχιστον 45:", count_at_least(extension_minutes, 45))
print("Κενή λίστα:", count_at_least([], 30))

Τουλάχιστον 30: 4
Τουλάχιστον 45: 2
Κενή λίστα: 0


## Πηγές αναφοράς
- [Python Tutorial · Αριθμοί, κείμενο και λίστες](https://docs.python.org/3/tutorial/introduction.html)
- [Python Tutorial · if, for, range και συναρτήσεις](https://docs.python.org/3/tutorial/controlflow.html)
- [Python · Ενσωματωμένες συναρτήσεις](https://docs.python.org/3/library/functions.html)
- [Google Colab · FAQ](https://research.google.com/colaboratory/faq.html)
- [ICO · Αποτελεσματικότητα ανωνυμοποίησης](https://ico.org.uk/for-organisations/uk-gdpr-guidance-and-resources/data-sharing/anonymisation/how-do-we-ensure-anonymisation-is-effective/)

Τα παραδείγματα και οι ασκήσεις δημιουργήθηκαν για το μάθημα. Δεν απαιτείται λήψη δεδομένων ή σύνδεση με εξωτερική υπηρεσία για την εκτέλεση του κώδικα.